File: Data Structures & Algorithms/same-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public static void dfs(boolean[] flag,TreeNode p,TreeNode q){
        if( p == null && q==null){
            return;
        }
        if( p == null &&  q!=null){
            flag[0]= false;
            return;
        }
        if( p!=null &&  q==null){
            flag[0]= false;
            return;
        }
        if( p.val !=  q.val){
            flag[0]= false;
            return;
        }
        dfs(flag,p.left,q.left);
        dfs(flag,p.right,q.right);
        return;

    }
    public boolean isSameTree(TreeNode p, TreeNode q) {
        boolean[] flag={true};
        dfs(flag,p,q);
        return flag[0];
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {

     function sametree(p,q){
        if((p===null) && (q===null)){
            return true
        }
        if(((p===null)&&(q!==null))|| ((p!==null)&&(q===null))){
            return false
        }
        if(p.val !== q.val){
            return false
        }
        return sametree(p.left,q.left) && sametree(p.right,q.right)
    }
     function backtrack(node){
        if(node===null){
            return false
        }
        if(sametree(node,subRoot)){
            return true
        }
        return backtrack(node.left) || backtrack(node.right)
     }
     return backtrack(root)
    }
}
